src/Stratis.Bitcoin/Base/TimeSyncBehavior.cs

```diff
@@ -296,6 +291,29 @@ private void RecalculateTimeOffsetLocked()
 
             this.logger.LogTrace("(-):{0}={1}", nameof(this.timeOffset), this.timeOffset.TotalSeconds);
         }
+         
+        /// <summary>
+        /// Gets the median offset of all the inbound samples 
+        /// limited to 1 less than the <see cref="OutboundToInboundWeightRatio"/> inbound per outbound
+        /// and the outbound samples with a weighting of the OutboundToInboundWeightRatio.
+        /// </summary>
+        /// <returns>Median offset in seconds.</returns>
+        private double GetMedianOffsetInSecondsFromInboundAndWeightedOutbound()
+        {
+            IEnumerable<double> inboundOffsets = this.inboundTimestampOffsets.Select(s => s.TimeOffset.TotalSeconds);
+            IEnumerable<double> outboundOffsets = this.outboundTimestampOffsets.Select(s => s.TimeOffset.TotalSeconds);
+
+            var allSamplesToConsider = new List<double>();
+
+            for (int i = 0; i < OutboundToInboundWeightRatio; i++)
+                allSamplesToConsider.AddRange(outboundOffsets);
+
+            int numberOfInboundSamplesToConsider = outboundOffsets.Count() * (OutboundToInboundWeightRatio - 1);
```

Comment:
why -1 here? 
You are basically changing the ratio from 1:10 to 1:9. 

So in case you have 1 outbound connection and a lot of inbound ones you will select 10 copies of outbound and first 9 inbound samples. 

So according to the comment 
 ```
       /// <summary>Ratio of weights of outbound timestamp offsets and inbound timestamp offsets.
        /// Ratio of N means that a single outbound sample has equal weight as N inbound samples.</summary>
        public const int OutboundToInboundWeightRatio = 10;
```
you should end up with 10 copies of that outbound one and first 10 inbound

Reply:
If the ratio change is justified then you should change the comment. But I don't think it is.

Reply:
Are you familiar with the issue that we are trying to solve here with this PR? #1035

Reply:
>Are you familiar with the issue that we are trying to solve here with this PR?

Indeed. 
The issue is that it is possible to get up to 200 potentially malicious samples while having a low amount of outbound samples. 
Old alghoritm trusts outbound samples 10 times more than inbound ones but because all inbound samples are used for the time adjustment it is still possible for an attacker to greatly overweight outbound samples. 

Your solution to this was to limit the amount of inbound samples so we are only considering `ratio * outboundCount` inbound samples. 

First problem with that solution is that you've proposed to change the ratio from 1:10 as it was before to 1:9. I don't see a justification for that. 

Secondly I don't think that we should not consider all inbound samples we have, instead we shjould rise the weight of the outbound ones. 


So the fix will be: 
never calculate adjusted time unless we have at least 1 outbound sample (that goes on top of that min 40 samples rule)

and change:

From

`
for (int i = 0; i < OutboundToInboundWeightRatio; i++)
    allSamples.AddRange(outboundOffsets);
`

To
 

```
int copiesCount = (int)((double)OutboundToInboundWeightRatio *
((double)this.inboundTimestampOffsets.Count) / (double)this.outboundTimestampOffsets.Count);

if (copiesCount <= 0) 
   copiesCount = 1;
for (int i = 0; i < copiesCount ; i++)
    allSamples.AddRange(outboundOffsets);
```

That way all inbound samples are always used and the 1:10 ratio is always in place.

Reply:
Now why I don't like not counting all the samples: more samples we have- more correct offset we get. 
Out of all inbound samples some can be malicious,  some are not. By using only the first few samples we potentially cutting out the good ones. 


Let's take a look at the possible attack scenario: attacker wants to change our offset as much as he can. When does he want to perform the attack? At the moment when we have the least amount of samples at possible- when we first appear online. So attacker will tend to have first few samples always reserved for him and in case a good node connects to us- we're not counting it since it's out of that range you propose.

That's why it's better to count all inbound samples, we just need to always keep the ratio of outbound to inbound ones at 1 : 10. Add enough copies of outbound samples until we reach that ratio and that's it.

Reply:
well you are right that instead of

```
int numberOfInboundSamplesToConsider = outboundOffsets.Count() * (OutboundToInboundWeightRatio - 1);
```

we could use 

```
int numberOfInboundSamplesToConsider = outboundOffsets.Count() * OutboundToInboundWeightRatio - 1;
```

but i believe both will produce same results most of the time

let's talk about your solution on skype

Reply:
ok, so now you two disagree on the implementation of reducing the inbound samples.

Personally, I am not familiar enough with what constitutes a geniune attack on the network either historically or logical potential.  

Looking at it from the outside - I would say that limiting to only 9 per outbound when the ratio is 10:1  appears to be an extra compensation..... 

>That's why it's better to count all inbound samples, we just need to always keep the ratio of outbound to inbound ones at 1 : 10. Add enough copies of outbound samples until we reach that ratio and that's it.

This suggestion..... would some example of that be.....

- 40 inbound,  outbound -> add 40 inbound and 40 copies of outbound
- 40 inbound, 2 outbound -> add 40 inbound and 20 of each outbound
-40 inbound, 3 outbound -> add 40 inbound, 13 of each outbound and perhaps 1 more of the first outbound (to make 40)

Is that what your suggested implementation is?

Reply:
Yep, cool

Reply:
By the way: we're perfectly safe even with the ratio of 1:3 instead of 1:10. Even of attacker can produce a small percentage of outbound samples (and by small I mean up to 25-30%).

Reply:
Here is the snippet to play around with values (create an empty console project and paste the snippet there):

```
class Program
    {
        public const int MaxTimeOffsetSeconds = 25 * 60;

        /// <summary>Ratio of weights of outbound timestamp offsets and inbound timestamp offsets.
        /// Ratio of N means that a single outbound sample has equal weight as N inbound samples.</summary>
        public const int OutboundToInboundWeightRatio = 3;

        private const int MinUnweightedSamplesCount = 40;


        static void Main(string[] args)
        {
            List<double> inboundOffsets = new List<double>();
            List<double> outboundOffsets = new List<double>() { 4,7,15,-12,6,2,-5, MaxTimeOffsetSeconds};


            for (int i = 0; i < 200; ++i)
            {
                inboundOffsets.Add(MaxTimeOffsetSeconds);
            }
            
            int unweightedSamplesCount = inboundOffsets.Count + OutboundToInboundWeightRatio * outboundOffsets.Count;

            if (unweightedSamplesCount >= MinUnweightedSamplesCount && outboundOffsets.Count > 0)
            {
                var median = GetMedianOffset(inboundOffsets, outboundOffsets);

                Console.WriteLine(median);
            }

            Console.WriteLine("FIN");
            Console.ReadLine();
        }

        private static double GetMedianOffset(List<double> inboundOffsets, List<double> outboundOffsets)
        {
            var allSamples = new List<double>();

            allSamples.AddRange(inboundOffsets);

            int copiesCount = (int)((double)OutboundToInboundWeightRatio * ((double)inboundOffsets.Count) / (double)outboundOffsets.Count);

            if (copiesCount <= 0)
                copiesCount = 1;
            for (int i = 0; i < copiesCount; i++)
                allSamples.AddRange(outboundOffsets);


            return allSamples.Median();
        }
    }

    public static class Ext
    {
        public static double Median(this IEnumerable<double> source)
        {
            int count = source.Count();
            if (count == 0)
                return 0;

            int midpoint = count / 2;
            IOrderedEnumerable<double> ordered = source.OrderBy(n => n);
            if ((count % 2) == 0)
                return ordered.Skip(midpoint - 1).Take(2).Average();
            else
                return ordered.ElementAt(midpoint);
        }
    }
```

Reply:
The weight ratio is not really a security here, so you are right that even another ratio will work. But we've selected 10 in order to say that for our those outbound values are much more valuable. So something like 5 or 10 would work, 3 not so much.

The security must be achieved using that mechanism that prevents inbound samples to take over outbound samples and also required attacker to have some outbound samples as well.

I think we are looking for something like requirement for attacker to have more than 25% or 33% of outbounds plus all inbounds

Reply:
>I think we are looking for something like requirement for attacker to have more than 25% or 33% of outbounds plus all inbounds

You mean less than 33%?

With max expected percentage of outbound samples that belong to the attacker being equal to 33% ratio of 1:4 is safe. 
Though we can go extra safe and have the 1:5, I don't mind. But 1:10 is overkill- with that ratio we kind of not considering inbound ones almost at all.

Reply:
I mean more, not less

this is not about the ratio between inbound and outbound, I think you have that incorrectly

the ratio then does not give you any protection, whatever the ratio is (unless something crazy)

if you consider a public node, it will likely have 8 outbound samples, but it will have a hundred of inbound samples. Therefore one to 10 ratio is not overkill, it reflects the default values for number of inbound and outbound connections.

So besides the ratio we then should define the security parameter saying how much of outbound connections we require attack to have in order to succeed with the attack. I think 25% or 33% could be a reasonable value, providing reasonable level  of security

Reply:
so imagine situation with samples
```
30 inbound
10x -5
10x -6
10x -7

4 outbound
1x 10
1x 11
1x 12
1x 13
```


with 1:10 weight ratio and median, we get result of `10`, So we see that if the attacker provided one outbound sample, the attack would be successful

with 1:100 weight ratio and median, we get result of `11`, So we see that if the attacker provided one outbound sample, the attack would be UNsuccessful, So we can see that the ratio of 100 provide protection against an attacker which is able to own 25% of outbound samples


now let's add 1000 inbound samples with value of -20 (suppose there is no limit on the above), this would put the median into the inbound sample territory, so even with 1:100 ratio, the attacker would not even need a single outbound sample

Therefore we need to limit the number of inbound samples, or as you suggested, dynamically compute the weight ratio for outbound samples

Reply:
>if you consider a public node, it will likely have 8 outbound samples, but it will have a hundred of inbound samples. Therefore one to 10 ratio is not overkill, it reflects the default values for number of inbound and outbound connections.

If we are talking about the original implementation- it's not a ratio, it's weight. It basically gives 10 times more trust to the outbound samples then to the inbound ones. 


Your and mine solutions are actually proposing to have a ratio. Which means it is guaranteed that for each inbound sample we are using 10 outbound ones. 
So for your or mine solution the ratio of 1:4 is sufficient. 
For old weighting solution weight multiplier of 10 is a good value (but since this solution is proven to have a flaw it is being changed in this PR).


>with 1:10 weight ratio and median, we get result of 10, So we see that if the attacker provided one outbound sample, the attack would be successful

Ah, I see. You are talking about old solution. My ration proposition is done under assumption that new logic is adopted. Your calculations are totally correct for the weight multiplayer. But since it's not the solution that will be implemented it's irrelevant.

Reply:
Have you read the last paragraph?

Reply:
>Have you read the last paragraph?

Yeah, it's also about good value for a weight multiplayer. 
I've mentioned that it's correct if we'd go for weight multiplayer.  
But I still don't understand why you are making calculations for the old solution instead of the new one. 

In terms of new one (one that guarantees that `inboundSamplesCount == outboundSamplesCount * ratio`) do you agree that the ratio value of 4 or 5 is safe?